src/app/rpmostree-libbuiltin.c

```diff
@@ -42,21 +42,47 @@ rpmostree_usage_error (GOptionContext  *context,
   (void) glnx_throw (error, "usage error: %s", message);
 }
 
+static char*
+get_id_from_deployment_variant (GVariant *deployment)
+{
+  g_autoptr(GVariantDict) dict = g_variant_dict_new (deployment);
+  g_autofree char *id;
+  g_assert (g_variant_dict_lookup (dict, "id", "s", &id));
+  return g_steal_pointer (&id);
+}
+
+static
+G_DEFINE_QUARK (rpmostree-original-id, rpmostree_original_id)
+#define RPMOSTREE_ORIGINAL_ID rpmostree_original_id_quark()
+
 static void
 default_deployment_change_cb (GObject *object,
                               GParamSpec *pspec,
-                              GVariant **value)
+                              gboolean   *changed)
 {
-  g_object_get (object, pspec->name, value, NULL);
+  GVariant *new_default_deployment;
+  g_object_get (object, pspec->name, &new_default_deployment, NULL);
+  g_autofree char *new_id = get_id_from_deployment_variant (new_default_deployment);
+
+  const char *original_id = g_object_get_qdata (object, RPMOSTREE_ORIGINAL_ID);
+  if (!g_str_equal (original_id, new_id))
+    *changed = TRUE;
 }
 
 void
 rpmostree_monitor_default_deployment_change (RPMOSTreeOS *os_proxy,
-                                             GVariant   **deployment)
+                                             gboolean    *changed)
 {
-  /* This will set the GVariant if the default deployment changes. */
+  g_autoptr(GVariant) default_deployment = rpmostree_os_dup_default_deployment (os_proxy);
+  g_autofree char *original_id = get_id_from_deployment_variant (default_deployment);
+
+  /* we use a quark here so original_id automatically gets freed with os_proxy (but also as
```

Comment:
I tend to view gobject data as a hack - it is a practical tool but the problem comes when you start using it a lot.  Which we aren't at all, this would be the first.  I'm OK keeping this but would be happier with a struct.  Though hum I forget now...is there some reason we can't just query the deployment property before and after?

Reply:
No reason, just wanted to make it easy for callers to use it without much prep. :)
Though it turned out pretty straightforward still. Fixup! ⬇️